### PR TITLE
Reshape maintanence

### LINKED ISSUE #197: Make Tensor.shape settable
```python
>>> x = mg.arange(6)
>>> x.shape = (2, 3)
>>> x
Tensor([[0, 1, 2],
        [3, 4, 5]])
```

### PR DESCRIPTION
**Pull request by rsokl**
Closes #197 

- Generalizes in-place operations on `Tensor`
- Fixes inconsistent interface between `Tensor.reshape` and `numpy.reshape`
- Fixes inconsistent param name between `mygrad.reshape` and `numpy.reshape`
- Makes `.shape`  settable
- Greatly improve tests for `reshape`
- Rename confusing internal param name to be more descriptive

### CODE REVIEW

File: src/mygrad/tensor_base.py [hunk 1]
```diff
@@ -782,6 +809,54 @@ def dtype(self):
...
+            inferred from the length of the tensor and remaining dimensions.
+
+        constant : bool, optional(default=False)
+            If ``True``, the returned tensor is a constant (it
+            does not back-propagate a gradient)
+
+        Returns
+        -------
+        mygrad.Tensor
+            ``a`` with its shape changed permuted.  A new tensor is returned.
```
Comment (davidmascharka) [inline src/mygrad/tensor_base.py]:
```suggestion
            ``a`` with its shape changed.  A new tensor is returned.
```

?

File: src/mygrad/tensor_base.py [hunk 2]
```diff
@@ -782,6 +809,54 @@ def dtype(self):
...
+        --------
+        >>> import mygrad as mg
+        >>> a = mg.Tensor([[1, 2, 3], [4, 5, 6]])
+        >>> a.reshape(6)
+        Tensor([1, 2, 3, 4, 5, 6])
+
+        >>> a.reshape(3, -1))   # the unspecified value is inferred to be 2
+        Tensor([[1, 2],
+                [3, 4],
+                [5, 6]])"""
```
Comment (davidmascharka) [inline src/mygrad/tensor_base.py]:
```suggestion
                [5, 6]])
                """
```

or do you want everything to end on the line?

File: src/mygrad/tensor_manip/array_shape/funcs.py [hunk 3]
```diff
@@ -1,59 +1,53 @@
...
+    --------
+    >>> import mygrad as mg
+    >>> a = mg.Tensor([[1,2,3], [4,5,6]])
+    >>> mg.reshape(a, 6)
+    Tensor([1, 2, 3, 4, 5, 6])
+
+    >>> mg.reshape(a, (3,-1))   # the unspecified value is inferred to be 2
+    Tensor([[1, 2],
+            [3, 4],
+            [5, 6]])"""
```
Comment (davidmascharka) [inline src/mygrad/tensor_manip/array_shape/funcs.py]: same here

File: src/mygrad/tensor_manip/array_shape/ops.py [hunk 4]
```diff
@@ -6,17 +6,17 @@
 
 
 class Reshape(Operation):
-    def __call__(self, a, shape):
+    def __call__(self, a, newshape):
         """ Parameters
             ----------
             a : mygrad.Tensor
-            shape : Tuple[int, ...]"""
+            newshape : Tuple[int, ...]"""
```
Comment (davidmascharka) [inline src/mygrad/tensor_manip/array_shape/ops.py]: same here

File: tests/custom_strategies/__init__.py [hunk 5]
```diff
@@ -313,3 +314,53 @@ def adv_integer_index(
...
         ),
     )
+
+
+@lru_cache(maxsize=1000)
+def _factors(n: int) -> List[int]:
+    """ Returns the divisors of n
+
+        >>> _factors(4)
+        {1, 2, 4}"""
```
Comment (davidmascharka) [inline tests/custom_strategies/__init__.py]: and here

Comment (rsokl) [reply tests/custom_strategies/__init__.py]: I'll do a sweep through the whole library and take care of it all at once

**Comment by rsokl** [comment]
@davidmascharka ugh I just realized that renaming `Tensor._op(constant)` to `Tensor._op(force_constant)` will break ops that you implemented in `mynn`. I guess I'll merge this into mygrad 2.0 instead

**Comment by rsokl** [comment]
closed in favor of #202

**Pull request closed without merging**